package/MDAnalysis/lib/_cutil.pyx

```diff
@@ -262,3 +264,78 @@ def make_whole(atomgroup, reference_atom=None):
         raise ValueError("AtomGroup was not contiguous from bonds, process failed")
     else:
         atomgroup.positions = newpos
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _bruteforce_capped(reference, configuration, max_cutoff, min_cutoff=None, box=None):
+    """Slightly optimised way of bruteforcing capped disdtance
+
+    In general does comparisons of squared distances to minimise calls to sqrt,
+    but still does len(a) * len(b) different comparisons
+    """
+    cdef bint min_crit, ortho, triclinic
+    cdef int Na, Nb, i, j, k
+    cdef float[:, :] a, b
+    cdef double vec_sq, max_sq, min_sq
+    cdef double vec[3]
+    cdef float[:] cbox
+    cdef float inverse_box[3]
+    cdef float tri_box[3][3]
+    cdef vector[int] indices  # will get reshaped to (n, 2)
+    cdef vector[float] distances
+
+    # to allow single coordinate inputs
+    if reference.shape == (3, ):
+        reference = reference[None, :]
+    if configuration.shape == (3, ):
+        configuration = configuration[None, :]
+    a = reference[:]
+    b = configuration[:]
+
+    min_crit = min_cutoff is not None
+    if min_crit:
+        min_sq = min_cutoff * min_cutoff
+    max_sq = max_cutoff * max_cutoff
+
+    Na = a.shape[0]
+    Nb = b.shape[0]
+
+    if box is not None:
+        cbox = box[:]
+        ortho = (box[3:] == 90.).all()
+        triclinic = not ortho
+
+        if ortho:
+            for i in range(3):
+                inverse_box[i] = 1.0 / cbox[i]
+        else:
+            from .mdamath import triclinic_vectors
+            tri_box = triclinic_vectors(box)
+    else:
+        ortho = False
+        triclinic = False
+
+    for i in range(Na):
+        for j in range(Nb):
+            for k in range(3):
+                vec[k] = a[i, k] - b[j, k]
+
+            # minimum image?
+            if ortho:
```

Comment:
Are if statements in inner loops time critical? Or would it better to have three code paths, one for each loop, one for ortho, for triclinic, and for no pbc?

Reply:
I guess it does not matter much for pbc because the cost of minimage is high. But for no-pbc it might make a difference (???).

Reply:
Branch prediction makes this not so bad for performance. See
https://stackoverflow.com/questions/11227809/why-is-it-faster-to-process-a-sorted-array-than-an-unsorted-array#11227902
for an explanation.

Reply:
Yeah it doesn't affect performance, I think Max is probably right as to why.  I think we're probably also memory bound rather than pure flop bound too, so a few bool comparisons are free